src/printer.js

```diff
@@ -3097,6 +3084,40 @@ function printFunctionParams(path, print, options, expandArg, printTypeParams) {
   ]);
 }
 
+function shouldPrintParamsWithoutParens(path, options) {
+  if (options.arrowFnParens === "always") {
+    return false;
+  }
+
+  const node = path.getValue();
+  if (options.arrowFnParens === "avoid") {
+    return canPrintParamsWithoutParens(node);
+  }
+
+  // "default" mode; omit parens only when:
+  // - currying (eg; fn = a => b => c)
+  // - used as a sole callback (eg; arr.map(x => x + 1))
```

Comment:
What about multiple callbacks?

e.g. you have this:

```js
promise.then(x => x + 1);
```

Then you add an error handler:

```js
promise.then(x => x + 1, error => log(error.toString()));
```

And this will result in:

```patch
-promise.then(x => x + 1, error => log(error.toString()));
+promise.then((x) => x + 1, (error) => log(error.toString()));
```

Which seems super strange to me.

Reply:
(I imagine you would use the `avoid` option)

This is intentional (though I agree it's tricky). 

It's easy to distinguish `(x =>` as a function, but slightly harder with `, x =>` (again, for folks with no functional background, who expect parens every time a function is declared or called). The parens also help make it clear that it is a second, independent function, and not related to the first arg (which otherwise can take some closer inspection vs. quick scanning). 

I started with just the rule, "only avoid parens for callbacks which are the first argument", but then you end up with `p.then(x => x, (e) => e)` which was too inconsistent for me to stomach, especially when both callbacks appear on their own line. 

I'm open to other ideas for this!